Add unit tests for audio processing utilities and content processing jobs

farmanahmed888 opened the pull request:
Introduce comprehensive unit tests for audio processing utilities and quiz content processing jobs to ensure robust input validation and error handling. These tests cover various scenarios, including invalid content structures and job lifecycle management.

farmanahmed888 committed:
Add unit tests for audio processing utilities and content processing jobs

farmanahmed888 committed:
Add unit tests for BlobService, TTS Service, and utility functions

farmanahmed888 committed:
Add comprehensive tests for teacher-webapp components and services

- Implement unit tests for App component, including rendering, form submission, and error handling.
- Create tests for callPage component to validate call management and participant interactions.
- Add tests for AddParticipantModal to ensure correct rendering and functionality.
- Introduce tests for StudentList and TeacherList components to verify selection behavior.
- Develop tests for ConferenceContext to check state management and event handling.
- Write integration tests for the full app flow, covering conference creation and error scenarios.
- Establish unit tests for apiService to validate API interactions and error handling.
- Create tests for state classes (Participant and AudioContentState) to ensure correct instantiation and properties.
- Mock global variables and functions for consistent testing environment.

farmanahmed888 committed:
feat(tests): add comprehensive test suite for WebSocket service

- Implemented Jest as the testing framework with scripts for running tests, coverage, and watch mode.
- Created a structured test directory with unit, integration, and performance tests.
- Added tests for application constants to ensure correct values and uniqueness.
- Developed performance tests for connection management and audio streaming efficiency.
- Created unit tests for Azure Blob service to validate blob data retrieval and error handling.
- Added unit tests for connection manager to verify CRUD operations and concurrent connections.
- Implemented control service tests to handle various message types and error scenarios.
- Developed WebSocket service tests to validate audio playback, state management, and error handling.
- Established a global test setup to mock console methods and environment variables.
- Created utility functions for mocking WebSocket connections and generating test data.

farmanahmed888 committed:
Add comprehensive testing strategies and test plans for ConferenceV2 and IVRv2

- Introduced a detailed testing strategy document for ConferenceV2, outlining unit, integration, and end-to-end testing approaches.
- Created a structured test directory for ConferenceV2, including unit tests for core services and integration tests for workflows.
- Developed a comprehensive test plan for IVRv2, detailing testing types, coverage areas, implementation plans, and quality metrics.
- Established guidelines for mocking external services and managing test data across both projects.
- Set performance and security testing strategies to ensure system reliability and robustness.

Soumabha commented:
A few action items here:
1. Add tests (at least one) for each of the other folders - teacher-webapp, IVRv2, Conferencev2, websocket-service, ContentWebApp and teacher-App.
2. Configure the execution of each of the folder based tests as part of the Github Actions with test coverage report being published as an artifact for view and evaluation. In case any of the tests fail, the pipeline should also fail.
3. Make appropriate modifications to the docs and README for the folders to represent how to run the tests (many of the folders already have it).

Also, currently there are multiple types of tests in backend-server, lets segregate them into unit and integration tests

- Unit tests: Validates only the code logic while mocking external service dependencies.
- Integration tests: Validates the code logic coupled with the external service dependency connections.

As of now, the new files are unit tests while the tenantRouter test is of integration nature which requires the DB_CONNECTION to be active.

Soumabha commented:
@farmanahmed888 there do not seem to be any changes committed post the last review

Soumabha commented:
@farmanahmed888 There are no updated tests available for the other repositories yet.

Soumabha commented on teacher-webapp/TEST_SUMMARY.md:
```diff
@@ -0,0 +1,147 @@
+# Test Suite Summary for Teacher WebApp
```
Please remove the Testing Summary, this is what is to be generated as an artifact as part of the CI/CD pipelines.

Soumabha reviewed (commented):
Lets finalise how we add testing docs within the repositories? 

Currently there are many different naming formats used -  TESTING_STRATEGY.md, test/README.md, TEST_PLAN.md, TESTING.md, TEST_SUMMARY.md

Soumabha commented:
Please add a test for the IvrV2 and ContentWebApp folders, the other folder tests seem fine.

farmanahmed888 committed:
refactor(tests): optimize mock data generation and improve performance test descriptions

farmanahmed888 committed:
Add unit tests for FSM, IVR, and utility functions

- Implement unit tests for the State and Transition classes in the FSM module.
- Create unit tests for the IVR functionality, including initialization and call handling.
- Add tests for utility functions, including HTML formatting and custom JSON encoding.
- Introduce tests for MongoDB utility class to ensure proper database interactions.
- Implement tests for SAS (Shared Access Signature) generator utility class.
- Update MongoDB connection handling to improve error checking for connection strings.

farmanahmed888 committed:
Merge branch 'a4i/main' into testing-pipeline-seeds

farmanahmed888 committed:
Refactor code structure for improved readability and maintainability

farmanahmed888 committed:
Add Azure deployment workflows for multiple services

farmanahmed888 committed:
Merge branch 'a4i/main' into testing-pipeline-seeds

farmanahmed888 committed:
Refactor deployment conditions in workflow files to remove unnecessary event checks

farmanahmed888 committed:
Refactor deployment conditions in main.yml to simplify workflow logic

farmanahmed888 committed:
Refactor quotes in main.yml for consistency and readability

farmanahmed888 committed:
Add render.yaml file with initial content

farmanahmed888 committed:
Update workflow files to include branch specification for push and pull_request events

farmanahmed888 committed:
Update workflow files to include testing-pipeline-seeds branch for push and pull_request events

farmanahmed888 committed:
Remove path specifications for push and pull_request events in main.yml

farmanahmed888 committed:
Create main workflow for WebSocket service

Set up CI/CD workflow for WebSocket service with testing and deployment to Docker, Render, and Azure.

farmanahmed888 committed:
Add workflows for backend server, conference v2, content web app, IVR v2, teacher web app, and websocket service; update render.yaml with backend server configuration

farmanahmed888 committed:
Update Docker metadata extraction to include PR event type and change SHA prefix

farmanahmed888 committed:
Remove obsolete workflows and consolidate deployment processes for Conference V2, Content Web App, IVR v2, Teacher Web App, and WebSocket Service. Update main workflow to call individual service workflows with improved Docker and Azure deployment configurations. Enhance Render deployment steps and streamline testing processes across services.

farmanahmed888 committed:
Update IVR v2 workflow to install test dependencies

farmanahmed888 committed:
Update Python version to 3.12 in IVR v2 and Conference V2 workflows

farmanahmed888 committed:
Update Docker metadata extraction to include PR event type and change SHA prefix

farmanahmed888 committed:
Use MemoryRouter in AddQuiz tests

Wrap AddQuiz component in MemoryRouter for routing context.

farmanahmed888 committed:
Fix rendering of AddQuiz component in tests

farmanahmed888 committed:
Update AddQuiz tests to use MemoryRouter

Wrap AddQuiz component in MemoryRouter for routing context in tests.

farmanahmed888 committed:
Enhance AddQuiz tests with MemoryRouter and improve test coverage; update setupTests to suppress React Router warnings

farmanahmed888 committed:
Update deployment scripts to accept HTTP 202 status code for successful triggers

farmanahmed888 committed:
Update Render deployment API key secret for GitHub Actions

farmanahmed888 committed:
Remove GitHub deployment flag from Render deployment step

farmanahmed888 committed:
Add Docker image verification step and update Render deployment configuration

farmanahmed888 committed:
Refactor deployment workflows to use specific service names and update Azure deployment configurations; remove obsolete render.yaml file

farmanahmed888 committed:
Update test scripts and add new tests for AddQuiz and AddContent components

farmanahmed888 committed:
Refactor and enhance testing suite for teacher-webapp

- Removed outdated state tests and migrated to new structure.
- Added comprehensive tests for callPage component, covering call management and participant interactions.
- Implemented tests for AddParticipantModal, ensuring proper selection and submission behavior.
- Created tests for StudentList and TeacherList components, validating rendering and selection logic.
- Developed integration tests for the full app flow, including conference creation and error handling.
- Enhanced apiService tests to cover all API interactions and error scenarios.
- Updated ConferenceContext tests to ensure correct state management and event handling.

farmanahmed888 committed:
Add missing ARG and ENV for REACT_APP_STORAGE_ACCOUNT_NAME in Dockerfile

farmanahmed888 committed:
Add new teacher and student entries to contacts.json

farmanahmed888 committed:
Update contacts.json to replace teacher and student entries

farmanahmed888 committed:
Update test data for teachers and students in App, integration, and state tests

farmanahmed888 committed:
Refactor integration and state tests by removing redundant tests and enhancing participant validation

farmanahmed888 committed:
Remove redundant selection state tests for StudentList and TeacherList components

Soumabha commented on IVRv2/requirements-test.txt:
```diff
@@ -0,0 +1,11 @@
+# Testing dependencies for IVR v2
```
Are we using separate requirements files for testing? Modern applications like Poetry have started using package managers to help differentiate these. You will be able to see a separate pattern in JS frameworks as well.


The better approach would be to contain test dependencies sep

farmanahmed888 replied:
done

Soumabha commented on ConferenceV2/services/singletons/conference_call_manager.py:
```diff
@@ -29,7 +29,7 @@ def __init__(
         self.ws_base_url = os.environ.get("WS_SERVER_EP", "")
 
-    def create_conference(self, teacher_phone: str, student_phones: List[str]) -> ConferenceCall:
+    def create_conference(self, teacher_phone: str, student_phones: List[str]) -> str:
```
It is a good practice to return the actual objects for usage with the UI. Make modifications in the UI code if required to accept the conference call JSON object.

farmanahmed888 replied:
done

Soumabha commented on teacher-webapp/src/contacts.json:
```diff
@@ -0,0 +1,51 @@
+{
```
Are these required for test or is it being used to display the teachers and the students on the UI?

If so, we should have it backend controlled and ensure that post tenant login, only teachers and students accessible by the tenants are visible.

farmanahmed888 replied:
Can't change the test structure for frontend react apps as [Link](https://github.com/facebook/create-react-app/releases/tag/v1.0.0)
says

> Ensure application and test files reside in src/
 We've never supported importing files from outside src/, nor have we supported running tests outside of src/.
We also never explicitly forbid doing so, which caused confusion when things didn't work like they should.


So we can't have uniform structure for test across front-end and back-end apps

Soumabha replied:
contacts.json refactor will be tackled with https://github.com/A4i-tech/.github/issues/61

For the tests folder , I think that documentation is their 1st release, as of now they support scanning anything which has a *.test.js at any depth of src. [Reading Link](https://create-react-app.dev/docs/running-tests/).

I tried it out locally for teacher-webapp and it did work, can you elaborate on the issue faced?

Soumabha commented:
Great job on structuring the pipeline for the monorepo with the main.yaml streamlining modification.

A few comments on the pipeline
1. The tests run with coverage but the coverage is not generated as an artifact for all the repositories
2. The test coverage should be updated in the individual README files to show up on top.

farmanahmed888 committed:
Update create_conference method to return ConferenceCall instance instead of conf_id

farmanahmed888 committed:
Refactor project dependencies: migrate from requirements.txt to pyproject.toml for ConferenceV2 and IVRv2

- Added pyproject.toml for ConferenceV2 with specified dependencies and testing tools.
- Removed requirements.txt for ConferenceV2.
- Added pyproject.toml for IVRv2 with specified dependencies and testing tools.
- Removed requirements-test.txt and requirements.txt for IVRv2.

farmanahmed888 committed:
Refactor testing documentation and structure across SEEDS platform

- Deleted outdated TEST_PLAN.md for IVRv2 and replaced it with a comprehensive unified test plan for the SEEDS platform, detailing testing strategies, coverage goals, and implementation guidelines.
- Removed specific testing documentation for Teacher WebApp and WebSocket Service, consolidating relevant information into the new unified test plan.
- Updated test structures and categories for ConferenceV2, IVRv2, Backend Server, WebSocket Service, and Teacher WebApp to ensure consistency and clarity.
- Enhanced coverage goals and performance benchmarks for all components, emphasizing the importance of integration and end-to-end testing.
- Established best practices for test organization, mocking strategies, and continuous integration to improve overall testing efficiency and reliability.

farmanahmed888 committed:
Fix create_start_conference and create_conference methods to return conf_id instead of conference_call_id

farmanahmed888 committed:
Enhance CI/CD workflows with coverage reporting and badges

- Updated conference-v2-main.yml to install Poetry, run tests with coverage, and upload coverage artifacts.
- Modified content-webapp-main.yml to run tests with coverage and upload coverage reports.
- Created coverage-summary.yml to aggregate coverage data from multiple services and comment on PRs.
- Enhanced ivr-v2-main.yml to utilize Poetry for dependency management and coverage reporting.
- Updated main.yml to include coverage summary job after all service workflows.
- Improved teacher-webapp-main.yml and websocket-service-main.yml to run tests with coverage and upload results.
- Added coverage badges to README files for all services.
- Introduced coverage setup documentation to guide users on configuring coverage reporting.
- Updated pyproject.toml for ConferenceV2 to include pytest-cov for coverage reporting.

farmanahmed888 committed:
Update coverage badge authentication tokens across multiple workflows and README files

farmanahmed888 committed:
Refactor Dockerfiles to use Poetry for dependency management instead of requirements.txt

farmanahmed888 committed:
Add caching for Python virtual environments and update dependencies for coverage reporting

farmanahmed888 committed:
Remove --no-dev option from Poetry install command in Dockerfiles

farmanahmed888 committed:
Refactor test commands in CI workflows to simplify coverage reporting and update Jest configuration for improved coverage metrics

farmanahmed888 committed:
Update .gitignore and configure Jest coverage reporters in multiple package.json files

farmanahmed888 committed:
Refactor coverage summary workflow to remove pull request trigger and streamline workflow calls

farmanahmed888 committed:
putting websocket on render

farmanahmed888 committed:
Merge branch 'a4i/main' into testing-pipeline-seeds

farmanahmed888 committed:
refactor monorepo

farmanahmed888 committed:
Update Dockerfile to use Node.js 20 and adjust .gitignore for coverage files

farmanahmed888 committed:
Refactor test commands to remove unnecessary flags and simplify coverage reporting

farmanahmed888 committed:
Enhance test script by adding flags to disable watch and progress output

farmanahmed888 committed:
Update test script to disable watch and progress output for cleaner test runs

farmanahmed888 committed:
Update coverage badge URL in README.md

farmanahmed888 committed:
Update coverage badge URL in README.md

farmanahmed888 committed:
Update coverage badge URLs in multiple README files
